### ISSUE TITLE
Is it possible to allow read access to the file being downloaded?

### ISSUE THREAD

**Issue by RationalFragile** [erickutcher/httpdownloader#221]
Hi, when I download a large file (say a movie) I want to be able to read the file (in VLC for example) while it's being downloaded.
I understand that I will only be able to play the first part being downloaded (since the other parts are not contiguous) but still, at least I can verify that the file is correct. Currently, files being downloaded (sparse files, no quick allocation) cannot be read without first stopping the download.
Thank you very much 😊

**Comment by erickutcher** [comment]
Sorry for such a late reply. I've tried implementing this because it's something that I've wanted to be able to do also, but I could never figure it out. It has to do with file access rights and probably other stuff. Maybe one of these days I'll solve it.

**Comment by RationalFragile** [comment]
@erickutcher, According to a stackoverflow, you need to remove other file permissions when creating the file: "Writer [processes] must use FILE_SHARE_READ, Reader [processes] must use FILE_SHARE_READ | FILE_SHARE_WRITE".

So just search for "FILE_SHARE_READ" and make sure other FILE_SHARE_X are removed. I don't know if it's necessary to remove FILE_SHARE_DELETE or not but you can just remove it since it's fine to require the user to stop the download before deleting the file.

Hope this helps 😊

**Comment by erickutcher** [comment]
Here's a beta version you can try out. I removed the DELETE and FILE_SHARE_DELETE flags from the CreateFile function. Those flags allowed the program to rename or delete a file while it's in use. I'll have to figure out a convenient way (probably an option) to allow both things.

[HTTP_Downloader_32_1.0.5.1_Beta.zip](https://github.com/erickutcher/httpdownloader/files/10777986/HTTP_Downloader_32_1.0.5.1_Beta.zip)
[HTTP_Downloader_64_1.0.5.1_Beta.zip](https://github.com/erickutcher/httpdownloader/files/10777987/HTTP_Downloader_64_1.0.5.1_Beta.zip)

**Comment by RationalFragile** [comment]
Thank youuu 😊 I can confirm that VLC can read the file while it's downloading now.

(Tho, I realized now that there is another small inconvenience for this use case: when using only one download stream (download parts), I think instead of using sparse files or dense files (whether zeroed or not), HTTP_Downloader should use normal "append" to file so that the video player (like VLC) would hit the end of the file and stop while waiting for the next frames instead of continuing to play the empty region.)

But, honestly, just sharing my idea, unless many other people want this, I think it's just not worth the hassle, especially dealing with when someone starts a single-stream download then switches it to multi-parts... And really, pausing VLC manually and waiting for the buffering a bit should be more than enough. And it's so much better than having to wait for a 24GB file to complete before playing it 😅

Again, thank you so much for your time and for the update.

**Comment by erickutcher** [comment]
I don't know if doing an append type download will have any effect. This link should download as a chunked transfer: https://filesamples.com/samples/video/mkv/sample_1280x720_surfing_with_audio.mkv

HTTP Downloader will download chunked transfers by appending to the end of the file. These types of downloads will show a "?" in the File Size column since their file size isn't known from the start.

VLC seems to bail when it hits a premature end of file.

**Comment by RationalFragile** [comment]
I see what you mean! Something fishy is going on. Because, I tried a few files and non of them made VLC wait.

I swear it happened before (the VLC bar shows an orange ping pong loading thing, while some other program like chrome is downloading it). But I can't now find any video to show this. And also this has nothing to do with HTTP_Downloader, I tested by downloading with Chrome and VLC still ended the playback instead of waiting.

So yeah, I was wrong in assuming that the file writing method was the one causing the issue.
